File: medium/0355_design_twitter/solution.cpp
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <vector>

class Twitter {
public:
  Twitter() {
    tweetsMap = std::unordered_map<int, std::vector<std::pair<int, int>>>();
    followsMap = std::unordered_map<int, std::unordered_set<int>>();

    timestamp = 0;
  }

  void postTweet(int userId, int tweetId) {
    tweetsMap[userId].push_back({timestamp++, tweetId});
  }

  std::vector<int> getNewsFeed(int userId) {
    std::priority_queue<std::tuple<int, int, size_t>> pq;

    if (tweetsMap.find(userId) != tweetsMap.end()) {
      const auto &tweets = tweetsMap[userId];
      pq.push({tweets.back().first, userId, tweets.size() - 1});
    }

    for (const int &followeeId : followsMap[userId]) {
      if (followeeId == userId)
        continue;

      if (tweetsMap.find(followeeId) != tweetsMap.end()) {
        const auto &tweets = tweetsMap[followeeId];
        pq.push({tweets.back().first, followeeId, tweets.size() - 1});
      }
    }

    std::vector<int> feed;

    while (feed.size() < 10 && !pq.empty()) {
      auto [_, userId, i] = pq.top();
      pq.pop();

      const auto &tweets = tweetsMap[userId];
      feed.push_back(tweets[i].second);

      if (i > 0)
        pq.push({tweets[i - 1].first, userId, i - 1});
    }

    return feed;
  }

  void follow(int followerId, int followeeId) {
    followsMap[followerId].insert(followeeId);
  }

  void unfollow(int followerId, int followeeId) {
    followsMap[followerId].erase(followeeId);
  }

private:
  std::unordered_map<int, std::vector<std::pair<int, int>>> tweetsMap;
  std::unordered_map<int, std::unordered_set<int>> followsMap;

  int timestamp;
};

File: medium/0355_design_twitter/solution.py
import heapq
from collections import defaultdict


class Twitter:
    def __init__(self):
        self.tweets = defaultdict(list)
        self.follows = defaultdict(set)
        self.time = 0

    def postTweet(self, user_id: int, tweet_id: int) -> None:
        self.tweets[user_id].append((-self.time, tweet_id))
        self.time += 1

    def getNewsFeed(self, user_id: int) -> list[int]:
        heap = []

        if self.tweets[user_id]:
            heap.append(
                (
                    *self.tweets[user_id][-1],
                    user_id,
                    len(self.tweets[user_id]) - 1,
                )
            )

        for followee_id in self.follows[user_id]:
            if followee_id == user_id:
                continue

            if self.tweets[followee_id]:
                heap.append(
                    (
                        *self.tweets[followee_id][-1],
                        followee_id,
                        len(self.tweets[followee_id]) - 1,
                    )
                )

        heapq.heapify(heap)
        news_feed = []

        while heap and len(news_feed) < 10:
            _, tweet_id, user_id, i = heapq.heappop(heap)
            news_feed.append(tweet_id)

            if i > 0:
                heapq.heappush(
                    heap, (*self.tweets[user_id][i - 1], user_id, i - 1)
                )

        return news_feed

    def follow(self, follower_id: int, followee_id: int) -> None:
        self.follows[follower_id].add(followee_id)

    def unfollow(self, follower_id: int, followee_id: int) -> None:
        self.follows[follower_id].discard(followee_id)
